src/terrain/terrain_helper.hpp

```diff
@@ -1,120 +1,152 @@
             for (int y = -1; y < 2; y++) {
-                if (x == 0 && y == 0) { continue; }
+                // for adjacent in same layer
+                if (x == 0 && y == 0) {
+                    continue;
+                }
                 // safety function to test if you xyz is in range;
-                if (!ter.in_range(tile->get_x() + x, tile->get_y() + y, tile->get_z())) { continue; }
+                if (!ter.in_range(
+                        tile->get_x() + x, tile->get_y() + y, tile->get_z()
+                    )) {
+                    continue;
+                }
                 Tile* adjacent_tile =
                     ter.get_tile(tile->get_x() + x, tile->get_y() + y, tile->get_z());
-                if (adjacent_tile->is_grass()
-                        & (getter(adjacent_tile)
-                            < ter.get_grass_grad_length() - 1)) {
-                    setter(adjacent_tile, ter.get_grass_grad_length() - 1);
+                // instead of height used maxheight +1
+                if (adjacent_tile->is_grass() && (getter(adjacent_tile) < height)) {
+                    // if adjacent tile is on the level below
+                    // it should be added to the next iteration
+                    next_grass_tiles.insert(adjacent_tile);
+                    // and the tiles with height less should have their height set
+                    setter(adjacent_tile, height);
                 }
             }
-            setter(tile, ter.get_grass_grad_length());
-            // tile->grow_sink should be set to true
-            next_grass_tiles.insert(tile);
-        }
     }
-    grow_grass_recursive<getter, setter>(ter, next_grass_tiles, ter.get_grass_grad_length() - 1);
+    grow_grass_inner<getter, setter>(ter, next_grass_tiles, height - 1);
 }
 
-// grow_grass_recursive(set, int height)
+// grow_grass recursive(set)
 // for tile in set
-//      if tile grow data high < height
-//          for adjacent tile (only if height != 0)
-//              if this tile is grass, and grow_data_high < height-1
-//                  add adjacent tiles to new set
-//          set tile grass height to height
-// if height then run grow_grass_recursive(new_set, height -1)
-template<int getter(Tile*), void setter(Tile*, int)>
-void grow_grass_recursive(Terrain& ter, std::set<Tile*> in_grass, int height){
-    if (height == 1) { return; }
+//      if tile is a source
+//          add adjacent tiles to the new set if they are grass and something about
+//          height set tile.grass height/source status
+// run grow_grass recursive(new_set, grass_grad_length)
+/*
+used to grow the first iteration of of grass
+given a set of all grass tiles gets the grass tiles that adjacent to an
+edge.
+Then next to those tiles set the grass height to max_hight-1.
+*/
+template <bool edge_detector(Tile*), int getter(Tile*), void setter(Tile*, int)>
+void
+grow_grass_recursive(Terrain& ter, std::set<Tile*> all_grass) {
```

Comment:
This has a lot of code duplication with the previous function

Reply:
no?
This function takes all grass, while the above function takes only a line of grass.
They cannot operate the same.

Reply:
Here's a diff without comments

![image](https://user-images.githubusercontent.com/76668284/198920853-c0c18695-35fc-4d96-bb3b-29d02d432d6a.png)

Reply:
that is literally three lines. the only reason that is long is because of the formatting.